README.md
WPHardening [![Build Status](https://travis-ci.org/elcodigok/wphardening.svg?branch=master)](https://travis-ci.org/elcodigok/wphardening) [![Coverage Status](https://coveralls.io/repos/elcodigok/wphardening/badge.svg?branch=master&service=github)](https://coveralls.io/github/elcodigok/wphardening?branch=master)
===========

<img src="https://raw.githubusercontent.com/elcodigok/wphardening/develop/doc/images/logo-wphardening-v1.png" alt="WPHardening" title="WPHardening" height="200px" align="right" />

Fortify the security of any WordPress installation.

**❮ NOTE ❯** This tool releases new versions on a regular basis. Make sure to update your dependencies frequently to get the latest version. Check out the [changelog](https://github.com/elcodigok/wphardening/releases) or [CHANGELOG.md](https://github.com/elcodigok/wphardening/blob/master/doc/CHANGELOG.md) to learn about the new features.

-----

Installation
------------

Installing WPHardening requires you to execute one console command:

```bash
$ pip install -r requirements.txt
```

Usage
-----

    $ python wphardening.py -h 

     __          _______  _    _               _            _
     \ \        / /  __ \| |  | |             | |          (_)
      \ \  /\  / /| |__) | |__| | __ _ _ __ __| | ___ _ __  _ _ __   __ _
       \ \/  \/ / |  ___/|  __  |/ _` | '__/ _` |/ _ \ '_ \| | '_ \ / _` |
        \  /\  /  | |    | |  | | (_| | | | (_| |  __/ | | | | | | | (_| |
         \/  \/   |_|    |_|  |_|\__,_|_|  \__,_|\___|_| |_|_|_| |_|\__, |
                                                                     __/ |
            Fortify the security of any WordPress installation.     |___/
    
         Caceria de Spammers - http://www.caceriadespammers.com.ar

    Usage: python wphardening.py [options]

    Options:
      --version             show program's version number and exit
      -h, --help            show this help message and exit
      -v, --verbose         Active verbose mode output results
      --update              Check for WPHardening latest stable version

      Target:
        This option must be specified to modify the package WordPress.

        -d DIRECTORY, --dir=DIRECTORY
                            **REQUIRED** - Working Directory.
        --load-conf=FILE    Load file configuration.

      Hardening:
        Different tools to hardening WordPress.

        -c, --chmod         Chmod 755 in directory and 644 in files.
        -r, --remove        Remove files and directory.
        -b, --robots        Create file robots.txt
        -f, --fingerprinting
                            Deleted fingerprinting WordPress.
        -t, --timthumb      Find the library TimThumb.
        --chown=user:group  Changing file and directory owner.
        --wp-config         Wizard generated wp-config.php
        --plugins           Download Plugins Security.
        --proxy=PROXY       Use a HTTP proxy to connect to the target url for
                            --plugins and --wp-config.
        --indexes           It deny you to display the contents of directories.
        --minify            Compressing static file .css and .js
        --malware-scan      Malware Scan in WordPress project.
        --6g-firewall       6G Firewall.
        --rest-api          Disable REST API.

      Miscellaneous:
        -o FILE, --output=FILE
                            Write log report to FILE.log


Examples
--------

### Check a WordPress Project

Before using the tool, we must ensure that our working directory is **WordPress**.

```bash
$ python wphardening.py -d /home/path/to/wordpress -v
```

### Change permissions

This option is to add the correct permissions to files and directories.

```bash
$ python wphardening.py -d /home/path/to/wordpress --chmod -v
```

### Remove files that are not used

Part of the fortification of any system is to remove those files, directories or components required.

```bash
$ python wphardening.py -d /home/path/to/wordpress --remove -v
```

### Create your robots.txt file

WordPress default does not incorporate the robots.txt file with this option poemos customize our robots.txt

```bash
$ python wphardening.py -d /home/path/to/wordpress --robots -v
```

For more information [robots.txt](http://www.robotstxt.org/)

### Remove all fingerprinting and Version

```bash
$ python wphardening.py -d /home/path/to/wordpress --fingerprinting -v
```

### Check a TimThumb library

```bash
$ python wphardening.py -d /home/path/to/wordpress --timthumb -v
```

### Create Index file

This file is created as a way to avoid sailing in a directory.

```bash
$ python wphardening.py -d /home/path/to/wordpress --indexes -v
```

### Download Plugins security

The following is a list of the most commonly used security plugins that you can download automatically:

 * [AntiVirus](https://wordpress.org/plugins/antivirus/)
 * [Bad Behavior](https://wordpress.org/plugins/bad-behavior/)
 * [Block Bad Queries](https://wordpress.org/plugins/block-bad-queries/)
 * [Exploit Scanner](https://wordpress.org/plugins/exploit-scanner/)
 * [Latch](https://wordpress.org/plugins/latch/)
 * [NinjaFirewall](https://wordpress.org/plugins/ninjafirewall/)
 * [Simple History](https://wordpress.org/plugins/simple-history/)
 * [Stream](https://wordpress.org/plugins/stream/)
 * [WP Security Scan](https://wordpress.org/plugins/wp-security-scan/)
 * [WP-DBManager](https://wordpress.org/plugins/wp-dbmanager/)
 * [WPS Hide Login](https://wordpress.org/plugins/wps-hide-login/)

```bash
$ python wphardening.py -d /home/path/to/wordpress --plugins
```

### Wizard generated wp-config.php

This command automatically creates a file called **wp-config-wphardening.php** which can then rename it.

```bash
$ python wphardening.py -d /home/path/to/wordpress --wp-config
```

### 6G Firewalls

```bash
$ python wphardening.py -d /home/path/to/wordpress --6g-firewall
```

### Disable REST API

```bash
$ python wphardening.py -d /home/path/to/wordpress --rest-api
```

### WPHardening update

With this option you can always have the latest version of WPHardening.

```bash
$ python wphardening.py --update
```

### Use all options

```bash
$ python wphardening.py -d /home/path/to/wordpress -c -r -f -t --wp-config --indexes --plugins --6g-firewall --rest-api -o /home/user/wphardening.log
```

Project Home
------------

www.caceriadespammers.com.ar


Git Repository
--------------

https://github.com/elcodigok/wphardening


Issues
------

https://github.com/elcodigok/wphardening/issues


doc/CHANGELOG.md
Changelog
=========


v1.6
----

### Featured

- Remove Theme twentyfifteen (FIX #82).
- Checking static files (FIX #83).
- Change url in Plugins, add https:// (FIX #87).
- Add Ninja Firewall Plugin for WordPress (FIX #81).
- Change permission 444 to wp-config-wphardening.php (FIX #95).
- Information Leak file error_log (FIX #97).
- Add 6G Firewall 2016 (FIX #88).
- Header X-Content-Type-Options (FIX #98).
- Security Headers - X-Frame-Options (FIX #100).
- Headers - X-XSS-Protection (FIX #99).
- Add WPS Hide Login Plugin for WordPress (FIX #107).
- Disable REST API (FIX #85).
- Summary of changes in HTML format (FIX #109).

### Improvement

- WPHardening compatible with WordPress 4.4, 4.4.1, 4.4.2, 4.5, 4.5.1, 4.5.2, 4.5.3, 4.6, 4.6.1, 4.7, 4.7.1, 4.7.2, 4.7.3 and 4.7.4
- Empty Trash 5 days (FIX #101).
- Disable Javascript Concatenation (FIX #105).
- Delete watermark on every files (FIX #92).


v1.5
----

### Featured

- Compressing static files *.css and *.js --minify (FIX #21).
- Changing file and directory Owner. --chown (FIX #50).
- Implement option WP_ALLOW_MULTISITE in wp-config.php (FIX #53).
- Remove Plugins readme.txt file (FIX #57).
- Enumeration Malware Scan (FIX #64).
- Disable automatic updates (FIX #66).
- Remove Theme Twentyfourteen (FIX #68).
- Stream recommended plugin (FIX #71).
- Simple History recommended plugin (FIX #72).
- Override of default file permissions in wp-config.php (FIX #73).
- Integration with Travis CI (FIX #77).
- Integration with Coveralls (FIX #78).

### Improvement

- Separate Class for menu (FIX #10).
- Improve malware scanner (FIX #49).
- Unify options --fingerprinting and --delete-version (FIX #51).
- Remove statics files (FIX #55).
- Protection wp-config.php file with .htaccess (FIX #65).
- Debug list of recommended plugins (FIX #67).
- Normalized Source Code to PEP8.
- WPHardening compatible with WordPress 4.1.1, 4.1.2, 4.2, 4.2.1, 4.2.2, 4.2.3, 4.2.4 and 4.3

### Bug

- Improve updateWPHardening class (FIX #38).
- Remove Full Path Disclosure (FIX #56).
- Debug data/wordpress.fuzz.txt list (FIX #61).
- Remove options php_flag display_errors off (FIX #79).


v1.4 (2014-11-29)
-----------------

- Add control Libwww-perl.
- Optimization source code.
- New set default vales in --wp-config
- Add Query string Exploits.
- Validation input options.
- Add error_reporting(0);
- Add @ini_set('display_errors', 0);
- Add filter login_error.
- Add Options Memori Limit in --wp-config
- Print version Timthumb library.
- Create new options --malware-scan
- Avoid Full Path Disclosure.
- New structure to eliminate themes.
- Create new options --load-conf


v1.3 (2014-07-31)
-----------------

- Change function wp_admin_css().
- Improved wordpress.fuzz.txt list.
- Handle the error when no Internet connection detected pluginsWordPress.py
- Handle the error when no Internet connection detected wpconfigWordPress.py
- Latch recommended plugin.
- Rename variable names.
- Improving detection of WordPress projects.
- Add auto-update core.
- Add a library to find timthumb.php
- Updated of PEP8 in the file wpconfigWordPress.py
- Updated of PEP8 in the file pluginsWordPress.py
- Full functionality with verbose mode.
- WPHardening compatible with WordPress 3.9, 3.9.1 and 3.9.2


v1.2 (2014-03-16)
-----------------

- New option in --wp-config to desctivar wp-cron.php
- Improvements in the new function to disable wp-cron.php
- Remove unsed libraries.
- New file extensions to modify fingerprintingWordPress.
- DISALLOW_UNFILTERED_HTML in wp-config-wphardening.php
- chmod 0640 to the file wp-config-wphardening.php
- FORCE_SSL_LOGIN and FORCE_SSL_ADMIN to the file wp-config-wphardening.php


v1.1 (2013-05-02)
-----------------

- License GPLv3 to all files in the source code.
- Protection for the .htaccess file.
- New record of activities in log file.
- New --wp-config option to generate the configuration file.
- Entering the URL option --robots


v1.0 (2013-04-12)
------------------

- New options --plugins
- New options -f --fingerprinting
- Create file robots.txt
- New options for changemod -c
- New options --indexes
- New options --delete-version
- Normalized PEP8.


lib/cmdLineParser.py
#!/usr/bin/env python
"""
cmdLineParser.py

Copyright 2013 Daniel Maldonado

This file is part of WPHardening project.

WPHardening is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

WPHardening is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with WPHardening.  If not, see <http://www.gnu.org/licenses/>.

"""


import os
from jinja2 import Environment, FileSystemLoader
import sys
import urllib2

from optparse import OptionParser
from optparse import OptionGroup

from lib.checkWordpress import checkWordpress
from lib.chmodWordPress import chmodWordPress
from lib.chownWordPress import chownWordPress
from lib.removeWordPress import removeWordPress
from lib.robotsWordPress import robotsWordPress
from lib.deleteVersionWordPress import deleteVersionWordPress
from lib.fingerprintingWordPress import fingerprintingWordPress
from lib.pluginsWordPress import pluginsWordPress
from lib.indexesWordPress import indexesWordPress
from lib.wpconfigWordPress import wpconfigWordPress
from lib.timthumbWordPress import timthumbWordPress
from lib.updateWPHardening import updateWPHardening
from lib.malwareScanWordPress import malwareScanWordPress
from lib.minifyWordPress import minifyWordPress
from lib.loadConfWordPress import loadConfWordPress
from lib.termcolor import colored
from lib.registerLog import registerLog
from lib.sixgWordPress import sixgWordPress
from lib.restApiWordPress import restApiWordPress


PATH = os.path.dirname(os.path.abspath(__file__))
TEMPLATE_ENVIRONMENT = Environment(
        autoescape=False,
        loader=FileSystemLoader(os.path.join(PATH, '../templates')),
        trim_blocks=False)


def render_template(template_filename, context):
    return TEMPLATE_ENVIRONMENT.get_template(template_filename).render(context)


def cmdBanner():
    """Banner printing."""
    print "\n"
    print """
     __          _______  _    _               _            _
     \ \        / /  __ \| |  | |             | |          (_)
      \ \  /\  / /| |__) | |__| | __ _ _ __ __| | ___ _ __  _ _ __   __ _
       \ \/  \/ / |  ___/|  __  |/ _` | '__/ _` |/ _ \ '_ \| | '_ \ / _` |
        \  /\  /  | |    | |  | | (_| | | | (_| |  __/ | | | | | | | (_| |
         \/  \/   |_|    |_|  |_|\__,_|_|  \__,_|\___|_| |_|_|_| |_|\__, |
                                                                     __/ |
             Fortify the security of any WordPress installation.    |___/

          Caceria de Spammers - http://www.caceriadespammers.com.ar
    """


def cmdLineParser():
    """Implementation to WPHardening."""

    version_wph = "1.6"
    usage = "usage: python %prog [options]"
    version = colored('WPHardening', 'green') + ' version ' + \
        colored(version_wph, 'yellow') + '\n'

    parser = OptionParser(usage, version=version)

    parser.add_option("-v", "--verbose", action="store_true",
                      dest="verbose", default=False,
                      help="Active verbose mode output results")

    parser.add_option("--update", action="store_true", dest="update",
                      default=False,
                      help="Check for WPHardening latest stable version")

    target = OptionGroup(parser, "Target", "This option must be "
                         "specified to modify the package WordPress.")

    target.add_option("-d", "--dir", dest="path", help="**REQUIRED** -"
                      " Working Directory.", metavar="DIRECTORY")

    target.add_option("--load-conf", dest="loadconf", metavar="FILE",
                      help="Load file configuration.")

    hardening = OptionGroup(parser, "Hardening", "Different tools to"
                            " hardening WordPress.")

    hardening.add_option("-c", "--chmod", action="store_true", dest="chmod",
                         help="Chmod 755 in directory and 644 in files.")

    hardening.add_option("-r", "--remove", action="store_true",
                         dest="remove", help="Remove files and directory.")

    hardening.add_option("-b", "--robots", action="store_true", dest="robots",
                         help="Create file robots.txt")

    hardening.add_option("-f", "--fingerprinting", action="store_true",
                         dest="finger", help="Deleted fingerprinting "
                         "WordPress.")

    hardening.add_option("-t", "--timthumb", action="store_true",
                         dest="timthumb", help="Find the library TimThumb.")

    hardening.add_option("--chown", action="store", type="string",
                         dest="chown", metavar="user:group", help="Changing "
                         "file and directory owner.")

    hardening.add_option("--wp-config", action="store_true", dest="wpconfig",
                         help="Wizard generated wp-config.php")

    hardening.add_option("--plugins", action="store_true", dest="plugins",
                         help="Download Plugins Security.")

    hardening.add_option("--proxy", action="store", type="string",
                         dest="proxy", help="Use a HTTP proxy to connect to "
                         "the target url for --plugins and --wp-config.")

    hardening.add_option("--indexes", action="store_true", dest="indexes",
                         help="It deny you to display the contents of "
                         "directories.")

    hardening.add_option("--minify", action="store_true", dest="minify",
                         help="Compressing static file .css and .js")

    hardening.add_option("--malware-scan", action="store_true",
                         dest="malwares", help="Malware Scan in WordPress "
                         "project.")

    hardening.add_option("--6g-firewall", action="store_true",
                         dest="sixg", help="6G Firewall.")

    hardening.add_option("--rest-api", action="store_true",
                         dest="api", help="Disable REST API.")

    miscellaneous = OptionGroup(parser, "Miscellaneous")

    miscellaneous.add_option("-o", "--output", help="Write log report to "
                             "FILE.log", metavar="FILE", dest="output")

    parser.add_option_group(target)
    parser.add_option_group(hardening)
    parser.add_option_group(miscellaneous)

    cmdBanner()

    (options, args) = parser.parse_args()

    if options.loadconf is not None:
        options.path = loadConfWordPress(options.loadconf).getDirectory()
        options.chmod = loadConfWordPress(options.loadconf).getChmod()
        options.robots = loadConfWordPress(options.loadconf).getRobots()
        options.finger = loadConfWordPress(
            options.loadconf
        ).getFingerprinting()
        options.wpconfig = loadConfWordPress(options.loadconf).getWpConfig()
        options.indexes = loadConfWordPress(options.loadconf).getIndexes()
        options.timthumb = loadConfWordPress(options.loadconf).getTimthumb()
        options.malwares = loadConfWordPress(options.loadconf).getMalwareScan()
        options.output = loadConfWordPress(options.loadconf).getOutput()

    if options.output is None:
        filename = 'wphardening.log'
    else:
        filename = options.output

    log = registerLog(filename)
    log.setConfigure()

    if options.update:
        log.add("Check for WPHardening latest stable version")
        updateWPHardening(os.path.abspath(".")).update()
        sys.exit()

    if options.path is None:
        log.add("Did not specify a working directory.")
        parser.print_help()
        sys.exit()

    options.path = os.path.abspath(options.path)

    if os.path.exists(options.path):

        fname = "output.html"
        context = {
            'directory': options.path,
            'version': version_wph
        }

        if checkWordpress(options.path, options.verbose).isWordPress():

            if options.chown is not None:
                changeOwner = chownWordPress(
                    options.path, options.chown, options.verbose
                )

                if changeOwner.isValid():
                    changeOwner.changeOwner()
                    context['chown'] = options.chown

            if options.chmod is not None:
                chmodWordPress(
                    options.path, options.verbose
                ).changePermisions()
                context['chmod'] = True

            if options.robots is not None:
                robotsWordPress(options.path).createRobots()
                context['robots'] = True

            if options.finger is not None:
                deleteVersionWordPress(options.path).delete()
                fingerprintingWordPress(
                    options.path, options.verbose
                ).searchStaticFile()
                context['finger'] = True

            if options.wpconfig is not None:

                if options.proxy is not None:
                    protocolo, rest = urllib2.splittype(options.proxy)

                    if protocolo is None:
                        raise ValueError("unknown URL type: %s") % \
                            (options.proxy)
                    host, rest = urllib2.splithost(rest)
                    host, port = urllib2.splitport(host)

                    if port is None:
                        raise ValueError("unknown protocol for %s") % \
                            (options.proxy)
                    puerto = int(port)
                    asdf = wpconfigWordPress(options.path, options.proxy)
                else:
                    asdf = wpconfigWordPress(options.path, options.proxy)
                asdf.createConfig()

            if options.indexes is not None:
                indexesWordPress(options.path, options.verbose).createIndexes()
                context['indexes'] = True

            if options.timthumb is not None:
                timthumbWordPress(options.path).checkTimbthumb()

            if options.plugins is not None:

                if options.proxy is not None:
                    protocolo, rest = urllib2.splittype(options.proxy)

                    if protocolo is None:
                        raise ValueError("unknown URL type: %s") % \
                            (options.proxy)
                    host, rest = urllib2.splithost(rest)
                    host, port = urllib2.splitport(host)

                    if port is None:
                        raise ValueError("unknown protocol for %s") % \
                            (options.proxy)
                    puerto = int(port)
                    asdf = pluginsWordPress(options.path, options.proxy)
                else:
                    asdf = pluginsWordPress(options.path, options.proxy)
                asdf.questions()

            if options.malwares is not None:
                malwareScanWordPress(options.path).scan()

            if options.remove is not None:
                removeWordPress(options.path).delete()
                context['remove'] = True

            if options.minify is not None:
                minifyWordPress(options.path, options.verbose).minify()

            if options.sixg is not None:
                sixgWordPress(options.path, options.verbose).createFirewall()
                context['sixg'] = True

            if options.api is not None:
                restApiWordPress(options.path).disableRestApi()
                context['api'] = True

        # output jinja2
        with open(fname, 'w') as f:
            html = render_template('index.html.tmpl', context)
            f.write(html)

    else:
        log.add("Could not find the specified directory.")
        print colored('\nCould not find the specified directory.\n', 'red')
